code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm

```diff
@@ -0,0 +1,748 @@
+	SLEEP_CHECK_DEATH(delay)
+	revving_charge = FALSE
+	var/movespeed = 0.8
+	walk_towards(src, T, movespeed)
+	SLEEP_CHECK_DEATH(get_dist(src, T) * movespeed)
+	walk(src, 0) // cancel the movement
+	charging = FALSE
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/MeleeAction(patience = TRUE)
+	if(charging)
+		return
+	return ..()
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/Bump(atom/A)
+	if(charging)
+		DestroySurroundings()
+		if(isliving(A))
+			var/mob/living/L = A
+			if(!istype(A, /mob/living/simple_animal/hostile/ancient_robot_leg))
+				L.visible_message("<span class='danger'>[src] slams into [L]!</span>", "<span class='userdanger'>[src] tramples you into the ground!</span>")
+				forceMove(get_turf(L))
+				var/limb_to_hit = L.get_organ(pick(BODY_ZONE_HEAD, BODY_ZONE_CHEST, BODY_ZONE_R_ARM, BODY_ZONE_L_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_LEG))
+				L.apply_damage(25, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration))
+				playsound(get_turf(L), 'sound/effects/meteorimpact.ogg', 100, TRUE)
+				shake_camera(L, 4, 3)
+				shake_camera(src, 2, 3)
+				if(mode == GRAV)
+					var/atom/throw_target = get_edge_target_turf(L, get_dir(src, get_step_away(L, src)))
+					L.throw_at(throw_target, 3, 2)
+	..()
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/proc/body_shield()
+	body_shield_enabled = TRUE
+	visible_message("<span class='danger'>[src] creates some sort of energy shield!</span>")
+	add_overlay("shield")
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/proc/disable_shield()
+	visible_message("<span class='danger'>[src]'s shield fails!</span>")
+	cut_overlay("shield")
+	body_shield_enabled = FALSE
+	addtimer(CALLBACK(src, .proc/body_shield), BODY_SHIELD_COOLDOWN_TIME)
+
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/bullet_act(obj/item/projectile/P)
+	if(body_shield_enabled)
+		do_sparks(2, 1, src)
+		visible_message("<span class='danger'>[src]'s shield deflects [P] in a shower of sparks!</span>", "<span class='userdanger'>You deflect the projectile!</span>")
+		if(P.damage)
+			disable_shield()
+		return
+	..()
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/attacked_by(obj/item/I, mob/living/user)
+	if(body_shield_enabled)
+		do_sparks(2, 1, src)
+		visible_message("<span class='danger'>[src]'s shield deflects [I] in a shower of sparks!</span>", "<span class='userdanger'>You deflect the attacl!</span>")
+		if(I.force)
+			disable_shield()
+		return
+	..()
+
+/mob/living/simple_animal/hostile/megafauna/ancient_robot/devour(mob/living/L)
+	say(pick("JKYZXAIZOBK GTGREYKX GIZOBK", "OTZKMXGZOTM YAHPKIZ YZXKTMZNY", "JKIUSVOROTM GTJ RKGXTOTM", "LOTJOTM IXOZOIGR CKGQTKYYKY")) //what can I say, I like the trope of something talking in cypher
+	visible_message("<span class='userdanger'>[src] disintigrates [L]!</span>","<span class='userdanger'>You analyse [L], restoring your health!</span>")
+	if(!is_station_level(z) || client)
```

Comment:
I would flip these checks. I am pretty sure checking for client first is more performant.

Reply:
can do